### PR TITLE
Helper methods to query flatfiles

### PR DESCRIPTION
**Pull request by carlacummins**
## Description

With the move from SQL tables to flatfiles, there is a greater need to query data in nested directories. Here, I've written some helper functions to aid in cases where we need to port an existing SQL query to work with flat files, e.g. in healthchecking/stats

**Related JIRA tickets:**
- ENSCOMPARASW-3023

#### New Bio::EnsEMBL::Compara::Utils::FlatFile methods
- 2 new methods have been included: one that handles the fetching of the data from the files and one that handles grouping the data in a way similar to SQL or Perl's DBI module (the first one calls the second).

## Testing
Unit test and accompanying test data is included. The tests pass.

## Notes
This PR is part 1 of 3-4 parts required to fulfill the JIRA ticket description. This is my attempt at breaking pull requests into smaller, more manageable units.

### CODE REVIEW

File: modules/Bio/EnsEMBL/Compara/Utils/FlatFile.pm [hunk 1]
```diff
@@ -121,4 +125,116 @@ sub match_range_filter {
     return $match;
 }
 
+=head2 query_file_tree
+
+  Arg [1]     : String $directory
+  Arg [2]     : String $file_extension
+  Arg [3]     : String or arrayref $selected_fields
+  Arg [4]     : String or arrayref $group_by
+  Description : Fetch data from data files under $directory. By deafult,
```
Comment (CristiGuijarro) [inline modules/Bio/EnsEMBL/Compara/Utils/FlatFile.pm:134]:
```suggestion
  Description : Fetch data from data files under $directory. By default,
```

File: modules/Bio/EnsEMBL/Compara/Utils/FlatFile.pm [hunk 2]
```diff
@@ -121,4 +125,116 @@ sub match_range_filter {
     return $match;
 }
 
+=head2 query_file_tree
+
+  Arg [1]     : String $directory
```
[the remark this answers is not in the document]
Comment (carlacummins) [reply modules/Bio/EnsEMBL/Compara/Utils/FlatFile.pm:130]: Looking through the codebase, it seems like `(optional)` is favoured over `(mandatory)`, so I will add this instead.

**Comment by JAlvarezJarreta** [comment]
`codecov` is definitely not working as expected. I will have a look at it tomorrow.

File: modules/t/Utils/FlatFile.t [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    # find absolute path to the test input
+    # important for travis-ci
+    use Cwd 'abs_path';
+    my $test_path = abs_path($0);
+    $test_path =~ s!FlatFile\.t!test_flatfiles/!;
+
+    my $full_data = [
+        { key1 => 0, key2 => 'a', key3 => 'z', key4 => 100 },
+        { key1 => 1, key2 => 'b', key3 => 'y', key4 => 100 },
+        { key1 => 2, key2 => 'd', key3 => 'v', key4 => 10  }, # this line comes from a *.tsv file - will be excluded later
```
Comment (JAlvarezJarreta) [inline modules/t/Utils/FlatFile.t:95]: I guess the comment could be removed now?
```suggestion
        { key1 => 2, key2 => 'd', key3 => 'v', key4 => 10  },
```

Comment (JAlvarezJarreta) [reply modules/t/Utils/FlatFile.t:95]: Oh no, I have misread the comment. Nevermind.

**Pull request merged**